src/main/java/com/gargoylesoftware/htmlunit/javascript/host/xml/XMLHttpRequest.java

```diff
@@ -165,101 +165,80 @@ public XMLHttpRequest(final boolean caseSensitiveProperties) {
         state_ = UNSENT;
     }
 
-    /**
-     * Returns the event handler that fires on every state change.
-     * @return the event handler that fires on every state change
-     */
-    @JsxGetter
-    public Function getOnreadystatechange() {
-        return stateChangeHandler_;
-    }
-
-    /**
-     * Sets the event handler that fires on every state change.
-     * @param stateChangeHandler the event handler that fires on every state change
-     */
-    @JsxSetter
-    public void setOnreadystatechange(final Function stateChangeHandler) {
-        stateChangeHandler_ = stateChangeHandler;
-        if (state_ == OPENED) {
-            setState(state_, null);
-        }
-    }
-
     /**
      * Sets the state as specified and invokes the state change handler if one has been set.
      * @param state the new state
      * @param context the context within which the state change handler is to be invoked;
      *                if {@code null}, the current thread's context is used.
      */
-    private void setState(final int state, Context context) {
+    private void setState(final int state, final Context context) {
         state_ = state;
 
-        final BrowserVersion browser = getBrowserVersion();
-        if (stateChangeHandler_ != null && (async_ || state == DONE)) {
-            final Scriptable scope = stateChangeHandler_.getParentScope();
-            final JavaScriptEngine jsEngine = (JavaScriptEngine) containingPage_.getWebClient().getJavaScriptEngine();
-
-            if (LOG.isDebugEnabled()) {
-                LOG.debug("Calling onreadystatechange handler for state " + state);
-
-                // do the decompilation an the debug output before the call, because the
-                // handler might deregister itself
-                if (context == null) {
-                    context = Context.getCurrentContext();
+        switch (state) {
```

Comment:
do we have to check before if the state is really different from the current one?

Reply:
In don't think so. In theory we could call LOADING multiple times to simulate the progress event.